File: src/llm/drivers/ollama/inference.rs
use async_trait::async_trait;
use reqwest::Client;

use crate::{
    http::{AppError, StandardErrorResponse},
    llm::{
        drivers::InferenceDriver,
        prompts,
        types::{InferenceOptions, InferenceRequest, InferenceResponseSync, InferenceStats},
    },
    utils,
};

use super::{
    types::{OllamaRequest, OllamaResponse},
    Ollama,
};

#[async_trait]
impl InferenceDriver for Ollama {
    async fn run_inference(
        &self,
        req: &InferenceRequest,
        options: &InferenceOptions,
    ) -> Result<InferenceResponseSync, AppError> {
        // TODO: We should default to using the chat completion endpoinnt. We can switch
        // to the generation endpoint only if the user explicitly specifies the prompt template/

        // Let's first get the prompt
        let prompt = prompts::get_prompt(&options.prompt_tmpl, &req.messages)?;

        // Prepare the request
        let req = OllamaRequest {
            model: &req.model,
            prompt: &prompt,
            raw: true,
            stream: Some(false),
            options: utils::merge_objects(
                &serde_json::json!({
                    "top_p": options.top_p,
                    "top_k": options.top_k,
                    "num_ctx": options.num_ctx,
                    "temperature": options.temperature,
                }),
                &options.driver_options,
            ),
        };

        // Call the Ollama API
        let client = Client::new();
        let url = format!("{}/api/generate", self.base_url);
        let response = client.post(&url).json(&req).send().await.map_err(|e| {
            eprintln!("Request error: {}", e);
            return AppError::BadRequest(StandardErrorResponse::new(
                "Unable to make request to Ollama".to_string(),
                e.to_string(),
            ));
        })?;

        let res = response.text().await.map_err(|e| {
            eprintln!("Response error: {}", e);
            return AppError::BadRequest(StandardErrorResponse::new(
                "Unable to get response from Ollama".to_string(),
                e.to_string(),
            ));
        })?;

        let mut res: OllamaResponse = serde_json::from_str(&res).map_err(|e| {
            eprintln!("Response parsing error: {}", e);
            return AppError::BadRequest(StandardErrorResponse::new(
                "Unable to parse Ollama response".to_string(),
                e.to_string(),
            ));
        })?;

        // Remove the weird ':' and '>' prefix that mistral keeps giving us.
        res.response = res.response.trim().to_string();
        if res.response.starts_with(':') || res.response.starts_with('>') {
            res.response = res.response[1..].to_string();
        }

        return Ok(InferenceResponseSync {
            model: res.model,
            created_at: res.created_at,
            response: res.response,
            stats: InferenceStats {
                total_duration: res.total_duration,
                load_duration: res.load_duration,
                prompt_eval_count: res.prompt_eval_count,
                prompt_eval_duration: res.prompt_eval_duration,
                eval_count: res.eval_count,
                eval_duration: res.eval_duration,
            },
            fn_call: None,
        });
    }
}

#[cfg(test)]
mod test {

    use crate::llm::types::InferenceMessage;

    use super::*;

    #[tokio::test]
    #[ignore]
    async fn test_ollama() {
        prompts::init(vec![
            crate::llm::prompts::PrompTemplate {
                name: "chatml".to_string(),
                tmpl: "{% for msg in messages %}<|im_start|>{{ msg.role }}\n{{ msg.content }}<|im_end|>\n{% endfor %}<|im_start|>assistant".to_string(),
                stop: vec!["<|im_start|>".to_string(), "<|im_end|>".to_string()],
            },
        ]);

        let driver = Ollama {
            base_url: "http://localhost:11434".to_string(),
        };

        let req = InferenceRequest {
            model: "mistral-openhermes".to_string(),
            messages: vec![
                InferenceMessage {
                    role: "system".to_string(),
                    content: Some("You are a helpful AI assistant".to_string()),
                },
                InferenceMessage {
                    role: "user".to_string(),
                    content: Some("What's the capital of British Columbia".to_string()),
                },
            ],
            tools: None,
        };

        let res = driver
            .run_inference(&req, &InferenceOptions::default())
            .await
            .unwrap();
        assert_eq!(res.model, "mistral-openhermes");
    }
}

File: modules/llm/types/inference.go
package types

import (
	"encoding/json"
	"time"
)

type InferenceRequest struct {
	Model        string             `json:"model"`
	Messages     []InferenceMessage `json:"messages"`
	Tools        []Tool             `json:"tools,omitempty"`
	OutputFormat *OutputFormat      `json:"format,omitempty"`
}

func NewInferenceRequest(model string, messages []InferenceMessage, tools []Tool) *InferenceRequest {
	return &InferenceRequest{
		Model:    model,
		Messages: messages,
		Tools:    tools,
	}
}

type OutputFormat string

const (
	OutputFormat_JSON OutputFormat = "json"
)

type InferenceMessage struct {
	Role    string `json:"role"`
	Content string `json:"content,omitempty"`
}

type Tool struct {
	Name        string          `json:"name"`
	Description string          `json:"description,omitempty"`
	Args        json.RawMessage `json:"args"`
	ToolType    ToolType        `json:"type"`
}

type ToolType string

const (
	Function ToolType = "function"
)

type InferenceOptions struct {
	TopP          *float64        `json:"top_p,omitempty"`
	TopK          *int32          `json:"top_k,omitempty"`
	NumCtx        *int32          `json:"num_ctx,omitempty"`
	Temperature   *float64        `json:"temperature,omitempty"`
	DriverOptions json.RawMessage `json:"driver_options"`
}

func NewInferenceOptions(topP *float64, topK *int32, numCtx *int32, temperature *float64) InferenceOptions {
	return InferenceOptions{
		TopP:          topP,
		TopK:          topK,
		NumCtx:        numCtx,
		Temperature:   temperature,
		DriverOptions: json.RawMessage(`{}`),
	}
}

func DefaultInferenceOptions() InferenceOptions {
	topP := 0.9
	topK := int32(40)
	numCtx := int32(4096)
	temperature := 0.2

	return InferenceOptions{
		TopP:          &topP,
		TopK:          &topK,
		NumCtx:        &numCtx,
		Temperature:   &temperature,
		DriverOptions: json.RawMessage(`{}`),
	}
}

type InferenceResponseSync struct {
	ID        string                   `json:"id"`
	Model     string                   `json:"model"`
	CreatedAt time.Time                `json:"created_at"`
	Response  InferenceResponseMessage `json:"response"`
	Stats     InferenceStats           `json:"stats"`
}

type InferenceResponseStream struct {
	Model        string                   `json:"model"`
	CreatedAt    time.Time                `json:"created_at"`
	Response     InferenceResponseMessage `json:"response"`
	FinishReason *FinishReason            `json:"finish_reason,omitempty"`
	Stats        *InferenceStats          `json:"stats,omitempty"`
}

type InferenceResponseMessage struct {
	Content      string        `json:"content"`
	FnCall       *FunctionCall `json:"fn_call,omitempty"`
	FinishReason FinishReason  `json:"finish_reason,omitempty"`
}

type FinishReason string

const (
	FinishReason_Stop          FinishReason = "stop"
	FinishReason_Length        FinishReason = "length"
	FinishReason_ToolCalls     FinishReason = "tool_calls"
	FinishReason_ContentFilter FinishReason = "content_filter"
	FinishReason_FunctionCall  FinishReason = "function_call"
)

type FunctionCall struct {
	Name       string          `json:"name"`
	Parameters json.RawMessage `json:"parameters"`
}

type InferenceStats struct {
	TotalDuration      uint64 `json:"total_duration,omitempty"`
	LoadDuration       uint64 `json:"load_duration,omitempty"`
	PromptEvalCount    uint64 `json:"prompt_eval_count,omitempty"`
	PromptEvalDuration uint64 `json:"prompt_eval_duration,omitempty"`
	EvalCount          uint64 `json:"eval_count,omitempty"`
	EvalDuration       uint64 `json:"eval_duration,omitempty"`
}
